add mutil-env docs (#233)
oam-dev/kubevela.io: docs/end-user/policies/envbinding.md

---

> title: Env-Binding

Reviewer: ```suggestion
title: Multi-Cluster Deploy
```

Reviewer 2: 这个文档应该叫 Multi-Environment Deployment

---
title: Env-Binding
---

> This documentation will introduce how to use env-binding to provide differentiated configuration and environment scheduling strategies for Applications.

Reviewer 2: ```suggestion
This documentation will introduce how to use env-binding to automate multi-stage application rollout across multiple environments.
```

## Background

> In daily development, users often deploy Application to different environments. For example, debug the Application in the development environment, and deploy Application in the production environment to provide services. For different environments, Application require differentiated configurations.

Reviewer 2: ```suggestion
Users usually have two or more environments to deploy applications to. For example, dev environment to test the application code, and production environment to deploy applications to serve live traffic. For different environments, the deployment configuration also has some nuance.
```

> ## How to

Reviewer 2: 标题直接讲要做的事情，比如 Multi-env application deployment

Author: 收到

Deploy the following Application:

```yaml
# app.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: policy-demo
  namespace: default
spec:
  components:
    - name: nginx-server
      type: webservice
      properties:
        image: nginx:1.21
        port: 80

  policies:
    - name: demo-policy
      type: env-binding
      properties:
        clusterManagementEngine: single-cluster
        envs:
          - name: test
            patch:
              components:
                - name: nginx-server
                  type: webservice
                  properties:
                    image: nginx:1.20
                    port: 80
            placement:
              namespaceSelector:
                name: test
                
    workflow:
      steps:
        - name: deploy-test-env
          type: multi-env
          properties:
            # Specify the policy name
            policy: demo-policy
            # Specify the env name in the policy
            env: test
```

> User wants to deploy the Application `policy-demo` to the namespace `test`(test environment) and modify some parameters in the component `nginx-server` to adapt to the test environment.

Reviewer 2: ```suggestion
This will deploy the Application `policy-demo` to the namespace `test`(test environment), and patch the parameters in the component `nginx-server` to adapt to the test environment.
```

> > policy `env-binding` needs to be used with workflow [multi-env](../workflow/multi-env) to complete application-level orchestration.

Reviewer 2: 去掉这行

Author: got it

> You need to provide several key parameters.

Reviewer 2: ```suggestion
The following config options are needed per environment:
```

Author: got it

- `patch` parameter is used to fill in differentiated configuration information. The `components` parameter in `patch` is a `Component` array, which is the same structure as the `Component` array in the Application. You need to pass the component `name` and `type` specify the component whose parameters you want to modify, and fill in the configured parameters in `properties`.

> - `placement` parameter is used to specify which cluster or namespace in the target environment to deploy the Application to. There are two kinds of selectors in `placement`: `clusterSelector` and `namespaceSelector`, respectively for the selection of cluster and namespace.

Reviewer 2: 上面这些字段讲解的内容不应该放在这里。直接放到后面单独成段解释。
这里主要给出 step by step walkthrough

Author: 明白，这些感觉可以都删掉

Next we apply the Application `policy-demo` in the example.

> Before applying an Application, you need a namespace named `test` in the current cluster. You can create it by executing cmd `kubectl create ns test`.

After the Application is created, a configured Application will be created under the `test` namespace.

```shell
$ kubectl get app -n test
NAME                           COMPONENT      TYPE         PHASE     HEALTHY   STATUS   AGE
demo-policy-test-policy-demo   nginx-server   webservice   running   true               65s
```

> If you want to use `env-binding` to create an Application in a multi-cluster environment, please refer to **[Multi-Cluster and Enviroment](../multi-app-env-cluster)**.

Reviewer 2: ```suggestion
If you want to learn more about `env-binding`, please refer to **[Multi-Cluster and Enviroment](../multi-app-env-cluster)**.
```
